src/devices/Arduino/ArduinoGpioControllerDriver.cs

```diff
@@ -0,0 +1,305 @@
+            finally
+            {
+                _arduinoBoard.Firmata.DigitalPortValueUpdated -= WaitForEventPortValueUpdated;
+            }
+
+            return new WaitForEventResult()
+            {
+                EventTypes = eventSeen,
+                TimedOut = false
+            };
+        }
+
+        protected override void AddCallbackForPinValueChangedEvent(int pinNumber, PinEventTypes eventTypes, PinChangeEventHandler callback)
+        {
+            lock (_callbackContainersLock)
+            {
+                if (_callbackContainers.TryGetValue(pinNumber, out var cb))
+                {
+                    cb.EventTypes = cb.EventTypes | eventTypes;
+                    cb.OnPinChanged += callback;
+                }
+                else
+                {
+                    var cb2 = new CallbackContainer(pinNumber, eventTypes);
+                    cb2.OnPinChanged += callback;
+                    _callbackContainers.Add(pinNumber, cb2);
+                }
+            }
+        }
+
+        protected override void RemoveCallbackForPinValueChangedEvent(int pinNumber, PinChangeEventHandler callback)
+        {
+            lock (_callbackContainersLock)
+            {
+                if (_callbackContainers.TryGetValue(pinNumber, out var cb))
+                {
+                    cb.OnPinChanged -= callback;
+                    if (cb.NoEventsConnected)
+                    {
+                        _callbackContainers.Remove(pinNumber);
+                    }
+                }
+            }
+        }
+
+        private void FirmataOnDigitalPortValueUpdated(object sender, PinValueChangedEventArgs e)
+        {
+            CallbackContainer? cb = null;
+            PinEventTypes eventTypeToFire = PinEventTypes.None;
+            lock (_callbackContainersLock)
+            {
+                if (_callbackContainers.TryGetValue(e.PinNumber, out cb))
+                {
+                    if (e.ChangeType == PinEventTypes.Rising && cb.EventTypes.HasFlag(PinEventTypes.Rising))
+                    {
+                        eventTypeToFire = PinEventTypes.Rising;
+                    }
+                    else if (e.ChangeType == PinEventTypes.Falling && cb.EventTypes.HasFlag(PinEventTypes.Falling))
+                    {
+                        eventTypeToFire = PinEventTypes.Falling;
+                    }
+                }
+            }
+
+            if (eventTypeToFire != PinEventTypes.None && cb != null)
+            {
+                cb.FireOnPinChanged(eventTypeToFire);
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                lock (_callbackContainersLock)
+                {
+                    _callbackContainers.Clear();
+                }
+
+                _arduinoBoard.Firmata.DigitalPortValueUpdated -= FirmataOnDigitalPortValueUpdated;
+            }
+
+            base.Dispose(disposing);
```

Comment:
consider adding null checks and nulling out all fields so GC can collect

Reply:
note: for nullable annotations just suppress messages with `!` since we do not care about correctness after dispose

Reply:
The fields are read-only, so I suppose a Clear() should mostly do it. Several of the values (i.e. the `_arduinoBoard` reference) aren't owned by this class, so nulling it has no effect anyway.